.github/workflows/ci.yaml

```diff
@@ -0,0 +1,491 @@
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# yamllint disable rule:line-length
+
+name: Continuous integration
+run-name: >-
+  Run CI checks for ${{github.event_name}} on ${{github.ref_name}}
+  by @${{github.actor}}
+
+on:
+  push:
+    branches:
+      - main
+
+  pull_request:
+    branches:
+      - main
+
+  merge_group:
+    types:
+      - checks_requested
+
+  workflow_dispatch:
+    inputs:
+      debug:
+        description: 'Run with debugging options'
+        type: boolean
+        default: true
+      soft-linting:
+        description: 'Do not quit for linting errors'
+        type: boolean
+        default: true
+
+env:
+  # Python version to use for actions/setup-python.
+  python-version: '3.13'
+  SHELLOPTS: ${{inputs.debug && 'xtrace'}}
+  PIP_PROGRESS_BAR: 'off'
+  # Global flag determining whether lint errors are fatal.
+  # TODO(mhucka): uncomment next line & delete 2nd after formatting PRs merged.
+  # soft-linting: ${{inputs.soft-linting}}
+  soft-linting: true
+
+concurrency:
+  cancel-in-progress: true
+  group: ${{github.workflow}}-${{github.event.pull_request.number||github.ref}}
+
+permissions: read-all
+
+jobs:
+  python-checks:
+    name: 'Python format & lint checks'
+    runs-on: ubuntu-slim
+    timeout-minutes: 20
+    steps:
+      - name: Check out a copy of the git repository
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+        with:
+          fetch-depth: 0
+
+      - name: Set up Python
+        uses: actions/setup-python@e797f83bcb11b83ae66e0230d6156d7c80228e7c # v5
+        with:
+          python-version: ${{env.python-version}}
+          cache: pip
+          cache-dependency-path: |
+            requirements.txt
+            dev-requirements.txt
+
+      - name: Install dependencies
+        run: pip install black~=25.9.0 flynt~=1.0 pylint~=4.0.0
```

Comment:
Please install `-r dev-requirements.txt` instead so you don't get out-of-sync versions here.  The python environment is cached just above so installing a few extra dependencies from dev-requirements should not matter.